core/src/services/gdrive/core.rs

```diff
@@ -0,0 +1,228 @@
+use std::fmt::Formatter;
+use std::sync::Arc;
+
+use crate::raw::HttpClient;
+use crate::Error;
+use crate::ErrorKind;
+use crate::Scheme;
+
+use http::request::Builder;
+use http::{header, Request, Response};
+use serde::Deserialize;
+use tokio::sync::Mutex;
+
+use crate::{
+    raw::{build_rooted_abs_path, new_request_build_error, AsyncBody, IncomingAsyncBody},
+    types::Result,
+};
+
+pub struct GdriveCore {
+    pub root: String,
+    pub access_token: String,
+    pub client: HttpClient,
+    pub path_cache: Arc<Mutex<HashMap<String, String>>>,
+}
+
+impl Debug for GdriveCore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut de = f.debug_struct("GdriveCore");
+        de.field("root", &self.root);
+        de.finish()
+    }
+}
+
+impl GdriveCore {
+    async fn get_abs_root_id(&self) -> Result<String> {
+        let root = "root";
+
+        if let Some(root_id) = self.path_cache.lock().await.get(root) {
+            return Ok(root_id.to_string());
+        }
+
+        let req = self
+            .sign(Request::get(
+                "https://www.googleapis.com/drive/v3/files/root",
+            ))
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        let resp = self.client.send(req).await?;
+        let resp_body = &resp.into_body().bytes().await.map_err(|e| {
+            Error::new(ErrorKind::Unexpected, "read response body error")
+                .with_context("service", Scheme::Gdrive)
+                .set_source(e)
+        })?;
+
+        let gdrive_file: GdriveFile = serde_json::from_slice(resp_body).map_err(|e| {
+            Error::new(ErrorKind::Unexpected, "deserialize json error")
+                .with_context("service", Scheme::Gdrive)
+                .with_context("origin json value", String::from_utf8_lossy(resp_body))
+                .set_source(e)
+        })?;
+
+        let root_id = gdrive_file.id;
+
+        let mut cache_guard = self.path_cache.lock().await;
+        cache_guard.insert(root.to_owned(), root_id.clone());
+
+        Ok(root_id)
+    }
+
+    async fn get_file_id_by_path(&self, file_path: &str) -> Result<String> {
+        let path = build_rooted_abs_path(&self.root, file_path);
+
+        if let Some(file_id) = self.path_cache.lock().await.get(&path) {
+            return Ok(file_id.to_string());
+        }
+
+        let mut parent_id = self.get_abs_root_id().await?;
+        let file_path_items: Vec<&str> = path.split('/').filter(|&x| !x.is_empty()).collect();
+
+        for (i, item) in file_path_items.iter().enumerate() {
+            let mut query = format!(
+                "name = '{}' and parents = '{}' and trashed = false",
+                item, parent_id
+            );
+            if i != file_path_items.len() - 1 {
+                query += "and mimeType = 'application/vnd.google-apps.folder'";
+            }
+            let query: String = query.chars().filter(|c| !c.is_whitespace()).collect();
```

Comment:
I didn't get this. `query` is built by our hand, why not build the query correct at the first place?

Reply:
query string can not contains white space. I keep the white space in query string for better readability and trim all white space after we built the query.

Reply:
You need to encode spaces in the querystring as `%20`, not to remove them. Because they are meaningful.

Reply:
how about `query = query.replace(" ", "%20");`?

Reply:
This is an operation with standards.

FYI: 
- https://github.com/apache/incubator-opendal/blob/30e82f5e3ebeea69938e8406fba5e71179c25538/core/src/raw/http_util/uri.rs#L46
- https://developer.mozilla.org/en-US/docs/Glossary/Percent-encoding